pkg/vere/io/lick.c

```diff
@@ -0,0 +1,645 @@
+  } 
+  else {
+    u3_noun dev, wir, cad, mar;
+    u3z(dat);
+    wir =  u3nc(c3__lick, u3_nul);
+    dev = _lick_string_to_path(gen_u->nam_c+1);
+    mar = u3i_string("error");
+    dat = u3i_string("not connected");
+    cad = u3nq(c3__soak, dev, mar, dat);
+    u3_auto_peer(
+      u3_auto_plan(&gen_u->lic_u->car_u, u3_ovum_init(0, c3__l, wir, cad)),
+      0, 0, 0);
+  }
+}
+
+
+/* _lick_io_kick(): apply effects.
+*/
+static c3_o
+_lick_io_kick(u3_auto* car_u, u3_noun wir, u3_noun cad)
+{
+  u3_lick* lic_u = (u3_lick*)car_u;
+
+  u3_noun tag, i_wir, par;
+  u3_noun nam, dat, tmp;
+  c3_o ret_o, dev_o;
+
+  if (  (c3n == u3r_cell(wir, &i_wir, 0))
+     || (c3n == u3r_cell(cad, &tag, &tmp))
+     || (c3__lick != i_wir) )
+  {
+    ret_o = c3n;
+  }
+  else {
+    if ( (c3__spin == tag) ) {
+      _lick_ef_spin(lic_u, u3k(tmp)); // execute spin command
+      ret_o = c3y;
+    } 
+    else if ( (c3__shut == tag) ) {
+      _lick_ef_shut(lic_u, u3k(tmp)); // execute shut command
+      ret_o = c3y;
+    } 
+    else if ( c3__spit == tag ) {
+      if ( c3y == u3r_cell(tmp, &nam, &dat) ) {
+        _lick_ef_spit(lic_u, u3k(nam), u3k(dat));
+      }
+      ret_o = c3y;
+    }
+    else {
+      ret_o = c3n;
+    }
+  }
+
+  u3z(wir);
+  u3z(cad);
+  return ret_o;
+}
+
+/* _lick_born_news(): initialization complete on %born.
+*/
+static void
+_lick_born_news(u3_ovum* egg_u, u3_ovum_news new_e)
+{
+  u3_auto* car_u = egg_u->car_u;
+
+  if ( u3_ovum_done == new_e ) {
+    car_u->liv_o = c3y;
+  }
+}
+
+/* _lick_born_bail(): %born is essential, retry failures.
+*/
+static void
+_lick_born_bail(u3_ovum* egg_u, u3_noun lud)
+{
+  u3l_log("lick: %%born failure;");
+  u3z(lud);
+  u3_ovum_free(egg_u);
+}
+
+/* _lick_io_talk(): notify %lick that we're live
+ *                  pass it a list of device names
+*/
+static void
+_lick_io_talk(u3_auto* car_u)
+{
+  u3_lick* lic_u = (u3_lick*)car_u;
+  u3_noun    wir = u3nc(c3__lick, u3_nul);
+  u3_noun    cad = u3nc(c3__born, u3_nul);
+
+  u3_auto_peer(
+    u3_auto_plan(car_u, u3_ovum_init(0, c3__l, wir, cad)),
+    0,
+    _lick_born_news,
+    _lick_born_bail);
+
+  car_u->liv_o = c3y;
+}
+
+/* _lick_io_exit(): close all open device files and exit
+*/
+static void
+_lick_io_exit(u3_auto* car_u)
+{
+  u3_lick* lic_u = (u3_lick*)car_u;
+  u3_port* cur_u = lic_u->gen_u;
+  u3_port* nex_u;
+  while ( NULL != cur_u ) {
+    _lick_close_sock(cur_u->san_u);
+    nex_u = cur_u->nex_u;
+    c3_free(cur_u);
+    cur_u = nex_u;
+  }
+
+  u3s_cue_xeno_done(lic_u->sil_u);
+  c3_free(lic_u);
+}
+
+
+/* u3_lick(): initialize hardware control vane.
+*/
+u3_auto*
+u3_lick_io_init(u3_pier* pir_u)
+{
+  u3_lick* lic_u = c3_calloc(sizeof(*lic_u));
+
+  c3_c pax_c[2048] = "";
+  struct stat st = {0};
+  strcat(pax_c, u3_Host.dir_c);
+  strcat(pax_c, URB_DEV_PATH);
+
+  if ( -1 == stat(pax_c, &st) ) {
+    u3l_log("lick init %s", lic_u->fod_c);
+    u3l_log("lick init mkdir %s",pax_c );
```

Comment:
We don't need two separate printfs here, just one. Also, `lic_u->fod_c` is null at this point.

Reply:
done